### PR TITLE
Migration/l10n br sale

### PR DESCRIPTION
**Pull request by renatonlima**

### CODE REVIEW

File: l10n_br_sale/models/sale.py [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+    def _fiscal_position_map(self, result, **kwargs):
+        ctx = dict(self.env.context)
+        kwargs['fiscal_category_id'] = ctx.get(
+            'fiscal_category_id')
+        ctx.update({'use_domain': ('use_sale', '=', True), 'fiscal_category_id': ctx.get(
+            'fiscal_category_id')})
+        return self.env['account.fiscal.position.rule'].with_context(
+            ctx).apply_fiscal_mapping(result, **kwargs)
+
+    #TODO - migrate to new api
```
Comment (danimaribeiro) [inline l10n_br_sale/models/sale.py]: https://github.com/odoo-brazil/l10n-brazil/issues/22
https://github.com/Trust-Code/l10n-brazil/commit/acb5fe74238cdca887e1362711d687b87bdb62a7

File: l10n_br_sale/models/res_config.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+from openerp import SUPERUSER_ID
+from openerp.osv import orm, fields
+
+
+class SaleConfiguration(orm.TransientModel):
+    _inherit = 'sale.config.settings'
+    _columns = {
+        'copy_note': fields.boolean(
+            u'Copiar Observações nos Documentos Fiscais'),
+    }
```
Comment (rvalyi) [inline l10n_br_sale/models/res_config.py]: would be cool to have that one in the v8 api. But as we are in a rush to get the OCA branch merged we can leave that to a next PR.

File: l10n_br_sale/models/sale.py [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+
+    @api.one
+    @api.depends('order_line.price_unit', 'order_line.tax_id',
+                 'order_line.discount', 'order_line.product_uom_qty')
+    def _amount_all_wrapper(self):
+        """ Wrapper because of direct method passing
+        as parameter for function fields """
+        return self._amount_all()
+
+    @api.one
```
Comment (rvalyi) [inline l10n_br_sale/models/sale.py:36]: :-1:  api.one is deprecated in 9. To help forward compatibility please use api.multi + ensure_one
see https://www.odoo.com/documentation/8.0/reference/orm.html#openerp.api.one

Better not introduce deprecated api here.

File: l10n_br_sale/models/sale.py [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+        result['fiscal_comment'] = " - ".join(fiscal_comment)
+        result['fiscal_category_id'] = fiscal_category_id
+
+        return result
+
+
+class SaleOrderLine(models.Model):
+    _inherit = 'sale.order.line'
+
+    @api.one
```
Comment (rvalyi) [inline l10n_br_sale/models/sale.py:246]: no api.one please

**Comment by mileo** [comment]
@sadamo hands on!

**Comment by rvalyi** [comment]
@mileo I didn't shout after this too much in the l10n_br_product_account module as it was blocking all other PR's. But as much as possible we should avoid them. The point is having the same code between v8 and v9 as much as possible. On these small modules I think it doesn't cost much to get it right at the merge already.

**Comment by mileo** [comment]
@rvalyi In v9 we have to pay attention in the option of changes quantities ( delivered and invoice ) in other documents. 

I do not see the code in details if we gonna have to override/deal with this features in l10n

**Comment by rvalyi** [comment]
@mileo yes there will be changes in these modules for sure. But still if we minimize the changes we maximize the ease to share the maintenance between v8 and v9; make it trivial to front port and backport fixes.

File: l10n_br_sale/models/sale.py (hunk 3 above)
Comment (rvalyi) [inline l10n_br_sale/models/sale.py:36]: forget these remarks: we are overriding super methods in these cases so we cannot change the api to multi. Furthermore, the purpose of using non deprecated new API is to have the closest code possible between 8 and 9. But in any case that will not be possible here as v9 comes with a cleaned new API in the sale order core.

File: l10n_br_sale/models/sale.py (hunk 4 above)
Comment (rvalyi) [inline l10n_br_sale/models/sale.py:246]: forget that, see https://github.com/OCA/l10n-brazil/pull/145#discussion_r45286291

**Comment by rvalyi** [comment]
ok so we are finally green on module installation + pep8.
This module being a glue module for sale_stock and sale_service, we could hardly test it standalone.
So tests will comes through the sale_stock and sale_service module themselves in PR;s coming next.
Still you can review this module meanwhile...

**Comment by rvalyi** [comment]
:+1:

**Comment by mileo** [comment]
:-1:  See Baeza commit on core https://github.com/odoo/odoo/commit/cbe85f42d067c48e1391ef567cee23fb6b27ed72

FIX here: https://github.com/akretion/l10n-brazil/pull/7

**Comment by rvalyi** [comment]
@mileo thanks for the head up and PR for increase modularity. I'll analyse that in a few hours.

**Comment by rvalyi** [comment]
@mileo that was trivial, I did the merge so it's now included in this PR.

**Comment by rvalyi** [comment]
@mileo I fixed a small bug in from your refactor. Now we are green on pep8/lint.

I made another temporary PR with l10n_br_sale_product https://github.com/OCA/l10n-brazil/pull/234 to show some test coverage on this l10_br_sale module:
https://coveralls.io/jobs/9844026

So this module is actually tested by downstream modules coming in (even more with stock and service).
So can we merge it now? (we have l10n_br_sale_product and l10n_br_stock coming in a matter of minutes too)

cc @renatonlima @danimaribeiro

**Comment by rvalyi** [comment]
(NB: coverage will be up again once we merge the downstream modules. The point is as this is really just a glue module testing it standalone is pointless as we already test its code with the downstream modules, as I have shown with the l10_br_sale_product temporary PR)

**Comment by renatonlima** [comment]
@rvalyi 

There is a little pep8 problem, but to me LGTM :+1:

**Comment by rvalyi** [comment]
@renatonlima the pep8 issue is in the other temporary branch to exibit the functional tests. The l10n_br_sale module is green on pep8 in this PR here.

**Comment by mileo** [comment]
I'll sleep a few hours, I'm basically up since yesterday and then begin to review the PR. 

@rvalyi @renatonlima

**Comment by rvalyi** [comment]
@mileo thanks for the incredible help. Been tough on our side too. Now I see we will have these PR's over this week finally. Fast tracking this merge now to unlock the next PR's, open for further fix anyway.

**Comment by mileo** [comment]
Back in game! Review in progress....

**Comment by mileo** [comment]
No review, lets change the procedures? If you will gonna do this for the next PR i will stop to review.
cc @danimaribeiro  @oca

**Comment by rvalyi** [comment]
@mileo sorry I thought it was ok as you just put a thumb up at https://github.com/OCA/l10n-brazil/pull/145#commitcomment-14514849
That's two with @renatonlima and that would be 3 if you would count me in.

Usually we would have to wait 5 open days, but you know that this project has very few active reviewers, that's why we aren't as strict during this convergence phase. The only other developer able to do constructive review in this project here is @danimaribeiro and he already signified several times publically he would prefer us to merge these PR's fast.

That being said if you exclude minor pep8 fixes, this PR has been around since June 9, so it's not like we would be taking everybody by surprise in a rush.

We will wait more for the next PR's I really thought you approved the merge. If you want to change something we will be reactive anyway.

**Comment by mileo** [comment]
I put a tumb in the fix that you do... "://github.com/OCA/l10n-brazil/pull/145#commitcomment-14514849"

For now please cherry-pick this commit:
1 - https://github.com/odoo-brazil/l10n-brazil/commit/5dba30e77f193db6729afd6c326801062835a684

**Comment by rvalyi** [comment]
@mileo yeah well now I got it sorry this wasn't obvious

**Comment by mileo** [comment]
It is not peeve, but have you noticed that only your point of view is correct?

**Comment by rvalyi** [comment]
@mileo sorry, thumb up after asking if we could merge was an approval for me. You said you would get some sleep I thought you were talking about the next PR's review...

**Comment by rvalyi** [comment]
@mileo I'm cherry-picking the commit. In the future we will ask for test for that kind of contrib (we didn't test it in 8.0 indeed). As for the past we all know the project has been maturing in sub-optimal conditions and no one is to blame for the lack of tests in v7 codebase. But in the future sticking to high coverage that kind of issue should disappear.

**Comment by rvalyi** [comment]
@mileo here you go https://github.com/OCA/l10n-brazil/pull/235

**Comment by mileo** [comment]
Can you create a new sale.order? I can´ t here.

2015-11-20 02:12:59,138 14194 ERROR sale_test2 openerp.http: Exception during JSON request handling.
Traceback (most recent call last):
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/openerp/http.py", line 537, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/openerp/http.py", line 574, in dispatch
    result = self._call_function(*_self.params)
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/openerp/http.py", line 310, in _call_function
    return checked_call(self.db, *args, *_kwargs)
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/openerp/service/model.py", line 118, in wrapper
    return f(dbname, _args, *_kwargs)
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/openerp/http.py", line 307, in checked_call
    return self.endpoint(_a, *_kw)
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/openerp/http.py", line 803, in __call__
    return self.method(_args, *_kw)
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/openerp/http.py", line 403, in response_wrap
    response = f(_args, *_kw)
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/addons/web/controllers/main.py", line 944, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/addons/web/controllers/main.py", line 936, in _call_kw
    return getattr(request.registry.get(model), method)(request.cr, request.uid, _args, *_kwargs)
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/openerp/api.py", line 250, in wrapper
    return old_api(self, _args, *_kwargs)
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/addons/sale/sale.py", line 368, in create
    new_id = super(sale_order, self).create(cr, uid, vals, context=ctx)
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/openerp/api.py", line 250, in wrapper
    return old_api(self, _args, *_kwargs)
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/addons/mail/mail_thread.py", line 381, in create
    thread_id = super(mail_thread, self).create(cr, uid, values, context=context)
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/openerp/api.py", line 250, in wrapper
    return old_api(self, _args, *_kwargs)
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/openerp/api.py", line 345, in old_api
    result = method(recs, _args, *_kwargs)
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/openerp/models.py", line 4092, in create
    record = self.browse(self._create(old_vals))
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/openerp/api.py", line 248, in wrapper
    return new_api(self, _args, *_kwargs)
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/openerp/api.py", line 481, in new_api
    result = method(self._model, cr, uid, _args, *_kwargs)
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/openerp/models.py", line 4298, in _create
    cr, user, ids, fields2, context)
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/openerp/api.py", line 250, in wrapper
    return old_api(self, _args, *_kwargs)
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/openerp/models.py", line 4436, in _store_set_values
    result = column.get(cr, self, ids, f, SUPERUSER_ID, context=context)
  File "/home/mileo/Projects/odoo/odoo8-oca/parts/server/openerp/osv/fields.py", line 231, in get
    raise Exception(_('undefined get method !'))
Exception: M\xe9todo get n\xe3o definido!

**Comment by mileo** [comment]
https://github.com/OCA/l10n-brazil/pull/236

**Comment by mileo** [comment]
The #235 avoid this error when invoicing with no admin user:

AccessError: ('AccessError', u'Voc\xea n\xe3o tem permiss\xe3o para criar este tipo de documento. Contacte o administrador do sistema se voc\xea acha que isto \xe9 um erro.\n\n(Modelo de documento: sale.advance.payment.inv)')

**Pull request merged**